### PR TITLE
feat!(api): GraphQL API key auth mode

### PR DESCRIPTION
**Pull request by Equartey**
*Description of changes:*
Adds API key auth mode to GraphQL Requests, error handling, and relevant unit tests.

Additionally, response errors are now nullable (breaking change).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/api/amplify_api/test/graphql_error_test.dart [hunk 1]
```diff
@@ -68,6 +68,6 @@ void main() {
         .response;
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/graphql_error_test.dart:68]: request: now that we are changing the expected logic for errors, could you please test that a happy case response has null errors? You could just add an assertion on one of the query tests or add another test.

File: packages/api/amplify_api/test/dart_graphql_test.dart [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+      final operation = Amplify.API.mutate(request: req);
+      final res = await operation.value;
+
+      final expected = json.encode(_expectedMutateSuccessResponseBody['data']);
+
+      expect(res.data, equals(expected));
+    });
+
+    test('Errors are decoded in response.errors', () async {
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/dart_graphql_test.dart:126]: request: please add a test that errors are null when the graphql response JSON from server has no errors.

File: packages/api/amplify_api/test/dart_graphql_test.dart [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+void main() {
+  group('GraphQL Dart API', () {
+    late AmplifyAPI dartApi;
+
+    setUpAll(() async {
+      await Amplify.addPlugin(MockAmplifyAPI());
+      await Amplify.configure(amplifyconfig);
+    });
+
+    test('Query returns proper response.data', () async {
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/dart_graphql_test.dart:86]: If you look at the tests for method channel implementation in amplify_api_query_test.dart I think there is more coverage that would be good to have parity with here. Some of that stuff might not make sense to test in dart implementation, like serialization/deserialization logic. However, I think we should at least test queries that use model helpers to ensure non-string requests are converted properly.

File: packages/api/amplify_api/test/dart_graphql_test.dart [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        } ''';
+      final req = GraphQLRequest(document: graphQLDocument, variables: {});
+
+      final operation = Amplify.API.query(request: req);
+      final res = await operation.value;
+
+      final expected = json.encode(_expectedQuerySuccessResponseBody['data']);
+
+      expect(res.data, equals(expected));
+    });
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/dart_graphql_test.dart:104]: Would it be possible to add a test in query and mutation for operation cancelation? You can look in the REST dart tests for an example. Arguably not necessary bc it's essentially testing that this is a cancelable operation but IMO good to have a check on that API surface.

File: packages/api/amplify_api/test/dart_graphql_test.dart [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    test('Mutate returns proper response.data', () async {
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/dart_graphql_test.dart:106]: same comment for queries applies to mutations. Could we check coverage in method channel implementation and ensure we have parity? Specifically with model helper mutations.

File: packages/api/amplify_api/lib/src/util.dart [hunk 6]
```diff
@@ -30,3 +30,21 @@ Map<String, String>? addContentTypeToHeaders(
...
+
+/// Grabs errors from GraphQL Response. Is used in method channel and Dart first code.
+/// TODO: Move to Dart first code when method channel GraphQL implementation is removed.
+@internal
+List<GraphQLResponseError> deserializeGraphQLResponseErrors(
+  Map<String, dynamic> response,
+) {
+  final errors = response['errors'] as List?;
+  if (errors == null || errors.isEmpty) {
+    return const [];
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/util.dart:42]: should this return null and then the return type of this method would be `List<GraphQLResponseError>?`

File: packages/api/amplify_api/test/dart_graphql_test.dart [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      'locations': _errorLocations,
+      'path': _errorPath,
+      'extensions': _errorExtensions,
+    },
+  ]
+};
+
+class MockAmplifyAPI extends AmplifyAPIDart {
+  @override
+  http.Client getGraphQLClient({String? apiName}) =>
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/dart_graphql_test.dart:61]: nice job with this mocking logic. I think this will change as more cases get covered, or get copied/extended but this is a nice start.

File: packages/api/amplify_api/test/dart_graphql_test.dart [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      MockClient((request) async {
+        if (request.body.contains('TestMutate')) {
+          return http.Response(
+              json.encode(_expectedMutateSuccessResponseBody), 400);
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/dart_graphql_test.dart:65]: nit/question: Does AppSync actually give a 400 status when there are errors like this? If not, would be good to match their server in the mock in case this had an impact on deserializing the server response (though I would not expect it to here).

File: packages/api/amplify_api/test/dart_graphql_test.dart [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        return http.Response(
+            json.encode(_expectedQuerySuccessResponseBody), 200);
+      });
+}
+
+void main() {
+  group('GraphQL Dart API', () {
+    late AmplifyAPI dartApi;
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/dart_graphql_test.dart:79]: nit/question: Is this needed? I think it may have been left in the dart REST test by mistake, in which case you could remove there and here.

File: packages/api/amplify_api/lib/src/amplify_api_config.dart [hunk 10]
```diff
@@ -29,7 +29,8 @@ class EndpointConfig with AWSEquatable<EndpointConfig> {
 
   /// Gets the host with environment path prefix from Amplify config and combines
   /// with [path] and [queryParameters] to return a full [Uri].
-  Uri getUri(String path, Map<String, dynamic>? queryParameters) {
+  Uri getUri(String? path, Map<String, dynamic>? queryParameters) {
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/amplify_api_config.dart:32]: request: Would it be possible to add unit test coverage for this class/utility here? I know it was not created in this PR but would be good to test the logic here regarding how it gets URIs from config, especially when inputs are null. Arguably this code could be simplified but would rather see a test, first.

File: packages/api/amplify_api/lib/src/api_plugin_impl.dart [hunk 11]
```diff
@@ -130,6 +152,30 @@ class AmplifyAPIDart extends AmplifyAPI {
     _authProviders[authProvider.type] = authProvider;
   }
 
+  // ====== GraphQL ======
+
+  @override
+  CancelableOperation<GraphQLResponse<T>> query<T>(
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/api_plugin_impl.dart:158]: I have an offline comment/question about this.

**Comment by ragingsquirrel3** [comment]
Would it be possible to change the beginning of the PR title to be "feat!(api): ...." to indicate that this has a breaking change? When you eventually merge, could you please write: "BREAKING CHANGE: graphql response errors now nullable" or something similar at the bottom of the commit message that gets used as the merge commit by gh during squash merge.

**Review by ragingsquirrel3 — CHANGES_REQUESTED** [review]

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import 'graphql_response_decoder.dart';
+
+@internal
+Future<GraphQLResponse<T>> GraphQLSendRequest<T>(
+    {required GraphQLRequest<T> request,
+    required http.Client client,
+    required Uri uri}) async {
+  final body = {'variables': request.variables, 'query': request.document};
+  final graphQLResponse = await client.post(uri, body: json.encode(body));
+
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:32]: I think there is room for improvement here regarding error handling/robustness. What if response is not a 200? What if JSON does not have expected structure? Would be good to have a look at other platforms and get an idea how this should work. Also suggest adding tests for that logic.

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    List<GraphQLResponseError> errors =
+        deserializeGraphQLResponseErrors(responseBody);
+
+    return GraphQLResponseDecoder.instance
+        .decode<T>(request: request, data: responseData, errors: errors);
+  }
+
+  // TEMP
+  throw Exception('Bad types');
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:45]: Going along with above comment about error handling, don't think we will throw this exception once that's refactored.

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+import 'package:meta/meta.dart';
+
+import '../util.dart';
+import 'graphql_response_decoder.dart';
+
+@internal
+Future<GraphQLResponse<T>> GraphQLSendRequest<T>(
+    {required GraphQLRequest<T> request,
+    required http.Client client,
+    required Uri uri}) async {
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:29]: nit: I think you can add a trailing comma here by putting a comma after `Uri uri` and then a newline after the comma. Eventually we need to remove all the info warnings from API category so it's easier to automatically enforce this.

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:http/http.dart' as http;
+import 'package:meta/meta.dart';
+
+import '../util.dart';
+import 'graphql_response_decoder.dart';
+
+@internal
+Future<GraphQLResponse<T>> GraphQLSendRequest<T>(
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:26]: nit: Should `graphQLSendRequest` not be capitalized here bc a function not a class?

File: packages/api/amplify_api/example/integration_test/graphql_tests.dart [hunk 16]
```diff
@@ -44,7 +44,7 @@ void main() {
       final req = GraphQLRequest<String>(
           document: graphQLDocument, variables: <String, String>{'id': id});
       final response = await Amplify.API.mutate(request: req).response;
-      if (response.errors.isNotEmpty) {
+      if (response.errors != null) {
```
Comment (dnys1) [inline packages/api/amplify_api/example/integration_test/graphql_tests.dart:47]: The [spec](https://spec.graphql.org/June2018/#sec-Errors) does say that this should either be `null` or non-empty, but we should still probably check if it's empty since we don't want to report an empty array as having error'd and AppSync doesn't always conform perfectly to the spec if I remember correctly.

```suggestion
      if (response.errors != null && response.errors!.isNotEmpty) {
```

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    required http.Client client,
+    required Uri uri}) async {
+  final body = {'variables': request.variables, 'query': request.document};
+  final graphQLResponse = await client.post(uri, body: json.encode(body));
+
+  final responseBody = json.decode(graphQLResponse.body);
+  if (responseBody is Map<String, dynamic>) {
+    final responseData = json.encode(responseBody['data']);
+
+    List<GraphQLResponseError> errors =
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:37]: Omitting types from local vars helps keep code consistent and concise and `final` is good documentation that the variable is not expected to change (vs. `var`)

```suggestion
    final errors =
```

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart (hunk 12 above)
Comment (dnys1) [reply packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:32]: The [spec](https://spec.graphql.org/June2018/#sec-Response-Format) implies that a 200 response guarantees a certain shape to the data. We can still verify that we get a Map back, but we must check if the status code is 200 first. Any other code means something went wrong.

File: packages/api/amplify_api/test/dart_graphql_test.dart (hunk 5 above)
Comment (Equartey) [reply packages/api/amplify_api/test/dart_graphql_test.dart:106]: Doesn't look like there are tests for mutations in `amplify_api_query_test.dart` currently.

File: packages/api/amplify_api/test/dart_graphql_test.dart (hunk 8 above)
Comment (Equartey) [reply packages/api/amplify_api/test/dart_graphql_test.dart:65]: Depending on the error, AppSync will return a 400, ie a bad GraphQL document will return a 400 with messaging. There is also no effect currently of the status code during decoding.

File: packages/api/amplify_api/example/integration_test/graphql_tests.dart [hunk 18]
```diff
@@ -561,7 +561,7 @@ void main() {
         // With stream established, exec callback with stream events.
         final subscription = await _getEstablishedSubscriptionOperation<T>(
             subscriptionRequest, (event) {
-          if (event.errors.isNotEmpty) {
+          if (event.errors != null) {
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/api/amplify_api/example/integration_test/graphql_tests.dart]: Good idea, added.

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+
+  final responseBody = json.decode(graphQLResponse.body);
+
+  if (responseBody is! Map<String, dynamic>) {
+    throw Exception('Response from GraphQL client incorrect type.');
+  }
+
+  // Handle Model responses that do not return a 'data' key
+  final responseData =
+      request.decodePath != null ? responseBody : responseBody['data'];
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:42]: Is this ternary necessary here? I ran the integ tests and saw lots of errors with the model helpers. I was thinking you would always want `responseBody['data']` because model helpers won't change the server response structure, just changes how the client decodes the response and doesn't require a change here. I think it was working before, just was suggesting adding some tests for it.

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:http/http.dart' as http;
+import 'package:meta/meta.dart';
+
+import '../util.dart';
+import 'graphql_response_decoder.dart';
+
+@internal
+Future<GraphQLResponse<T>> graphQLSendRequest<T>({
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:26]: nit: could you please add a little doc comment for this function saying what it does. I think this file will have no lint warnings after that. maybe something like:

```
/// Converts the [GraphQLRequest] to an HTTP POST request and sends with ///[client]. 
```

File: packages/api/amplify_api/example/integration_test/graphql_tests.dart [hunk 21]
```diff
@@ -561,7 +561,7 @@ void main() {
         // With stream established, exec callback with stream events.
         final subscription = await _getEstablishedSubscriptionOperation<T>(
             subscriptionRequest, (event) {
-          if (event.errors.isNotEmpty) {
+          if (event.hasErrors) {
             fail('subscription errors: ${event.errors}');
           }
           dataCompleter.complete(event);
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/example/integration_test/graphql_tests.dart:567]: request: Could we comment out or have an early return in the group below about subscriptions along with a TODO comment to re-enable them once subscriptions implemented. I would prefer to have failing integ test suites and I think all these should pass here (except subscriptions), which is awesome!

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+  required http.Client client,
+  required Uri uri,
+}) async {
+  final body = {'variables': request.variables, 'query': request.document};
+  final graphQLResponse = await client.post(uri, body: json.encode(body));
+
+  final responseBody = json.decode(graphQLResponse.body);
+
+  if (responseBody is! Map<String, dynamic>) {
+    throw Exception('Response from GraphQL client incorrect type.');
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:37]: nit: Could this be `ApiException` and change the message to something like "unable to parse GraphQLResponse from server response which was not a JSON object." Just thinking it might be more helpful to folks than something about type errors.

Comment (ragingsquirrel3) [reply packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:37]: Also, for the underlying error argument to `ApiException`, would it make sense to include the status code and the response body? If someone had an error here, they might have no way of seeing the server error unless we surface along with this error.

File: packages/api/amplify_api/test/dart_graphql_test.dart (hunk 8 above)
Comment (ragingsquirrel3) [reply packages/api/amplify_api/test/dart_graphql_test.dart:86]: Thanks for checking. I also saw that non-200 responses still included valid gql responses with the errors formatted correctly.

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart (hunk 22 above)
Comment (ragingsquirrel3) [reply packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:37]: It might also make sense to wrap the `json.decode(` in a try/catch as I would expect that to throw an error if the body wasn't JSON-able.

File: packages/api/amplify_api/lib/src/amplify_authorization_rest_client.dart [hunk 23]
```diff
@@ -50,8 +52,12 @@ class AmplifyAuthorizationRestClient extends http.BaseClient
...
     if (!request.headers.containsKey(AWSHeaders.authorization) &&
         endpointConfig.authorizationType != APIAuthorizationType.none) {
-      // ignore: todo
-      // TODO: Use auth providers from core to transform the request.
+      // TODO(ragingsquirrel3): Use auth providers from core to transform the request.
+      final apiKey = endpointConfig.apiKey;
+      if (endpointConfig.authorizationType == APIAuthorizationType.apiKey &&
+          apiKey != null) {
+        request.headers.putIfAbsent(_xApiKey, () => apiKey);
+      }
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/amplify_authorization_rest_client.dart:60]: I would rearrange this a bit. If the authorization type is API key but there is no API key we should alert the user (throw an error)

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+import 'graphql_response_decoder.dart';
+
+@internal
+Future<GraphQLResponse<T>> graphQLSendRequest<T>({
+  required GraphQLRequest<T> request,
+  required http.Client client,
+  required Uri uri,
+}) async {
+  final body = {'variables': request.variables, 'query': request.document};
+  final graphQLResponse = await client.post(uri, body: json.encode(body));
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:32]:
```suggestion
  final graphQLResponse = await client.post(uri, body: json.encode(body));
  
  if (graphQLResponse.statusCode != 200) {
    throw ApiException(...);
  }
```

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart (hunk 20 above)
Comment (dnys1) [reply packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:26]: +1. I would also rename it to `sendGraphQLRequest` per Dart [naming conventions](https://dart.dev/guides/language/effective-dart/design#prefer-an-imperative-verb-phrase-for-a-function-or-method-whose-main-purpose-is-a-side-effect).

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart (hunk 24 above)
Comment (Equartey) [reply packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart:32]: I'm not sure we want to short circuit here on non 200 codes. App sync returns valuable error message information despite a non successful status code (ie 400) and the process to decode those errors are what follows this.

File: packages/api/amplify_api/test/amplify_api_config_test.dart [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+
+void main() {
+  late EndpointConfig endpointConfig;
+
+  group('GraphQL Config', () {
+    const endpointType = EndpointType.graphQL;
+    const endpoint =
+        'https=//abc123.appsync-api.us-east-1.amazonaws.com/graphql';
+    const region = 'us-east-1';
+    const authorizationType = APIAuthorizationType.apiKey;
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/amplify_api_config_test.dart:17]: nit: if mocking api key type might as well add a fake api key here in case there was some code path that assumed those things connected (though that's not the case here).

File: packages/api/amplify_api/test/amplify_api_config_test.dart [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+      const config = AWSApiConfig(
+          endpointType: endpointType,
+          endpoint: endpoint,
+          region: region,
+          authorizationType: authorizationType);
+
+      endpointConfig = const EndpointConfig('GraphQL', config);
+    });
+
+    test('should return valid URI wtih null params', () async {
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/amplify_api_config_test.dart:29]: nit, spelling

File: packages/api/amplify_api/test/amplify_api_config_test.dart [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+      final uri = endpointConfig.getUri(path: null, queryParameters: null);
+      final expected = Uri.parse('$endpoint/');
+
+      expect(uri, equals(expected));
+    });
+
+    test('should returns invalid URI wtih params', () async {
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/amplify_api_config_test.dart:36]: here, too

File: packages/api/amplify_api/test/amplify_api_config_test.dart [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+      expect(uri, isNot(expected));
+    });
+  });
+
+  group('REST Config', () {
+    const endpointType = EndpointType.rest;
+    const endpoint =
+        'https=//abc123.appsync-api.us-east-1.amazonaws.com/graphql';
+    const region = 'us-east-1';
+    const authorizationType = APIAuthorizationType.apiKey;
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/amplify_api_config_test.dart:50]: nit: please do IAM w REST for this mock as that's the only thing the CLI will do for you at the moment and api key not possible w REST.

File: packages/api/amplify_api/test/amplify_api_config_test.dart [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+          endpointType: endpointType,
+          endpoint: endpoint,
+          region: region,
+          authorizationType: authorizationType);
+
+      endpointConfig = const EndpointConfig('REST', config);
+    });
+
+    test('should return valid URI wtih params', () async {
+      final path = 'path/to/nowhere';
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/amplify_api_config_test.dart:63]: request: please add a test where the path has a leading slash like '/path' bc that's something we account for

File: packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart (hunk 19 above)
Comment (Equartey) [reply packages/api/amplify_api/lib/src/graphql/graphql_send_request.dart]: Resolved offline. Ternary is not needed.

File: packages/api/amplify_api/test/amplify_api_config_test.dart [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+
+import 'test_data/fake_amplify_configuration.dart';
+
+void main() {
+  late EndpointConfig endpointConfig;
+
+  group('GraphQL Config', () {
+    const endpointType = EndpointType.graphQL;
+    const endpoint =
+        'https=//abc123.appsync-api.us-east-1.amazonaws.com/graphql';
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/amplify_api_config_test.dart:15]: nit: could you please make the first part "https://" so it's more like a normal URI instead of with "="?

File: packages/api/amplify_api/test/amplify_api_config_test.dart [hunk 31]
```diff
@@ -0,0 +1,84 @@
...
+      final params = {'foo': 'bar', 'bar': 'baz'};
+      final uri = endpointConfig.getUri(path: path, queryParameters: params);
+
+      final expected = Uri.parse('$endpoint/$path?foo=bar&bar=baz');
+
+      expect(uri, equals(expected));
+    });
+
+    test('should handle a leading slash', () async {
+      final path = 'path/to/nowhere/';
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/amplify_api_config_test.dart:75]: request: for leading slash, could this be '/path/to/nowwhere' ?

File: packages/api/amplify_api/test/amplify_api_config_test.dart [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+    });
+
+    test('should return valid URI with null params', () async {
+      final uri = endpointConfig.getUri(path: null, queryParameters: null);
+      final expected = Uri.parse('$endpoint/');
+
+      expect(uri, equals(expected));
+    });
+
+    test('should returns invalid URI with params', () async {
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/amplify_api_config_test.dart:38]: question: What is this testing and why is it invalid? Seems like this logic is tested below (although for REST) so not sure this test is necessary. If you do keep it, nit on grammar of description "should return ...."

File: packages/api/amplify_api/lib/src/graphql/send_graphql_request.dart [hunk 33]
```diff
@@ -0,0 +1,57 @@
...
+    if (responseBody is! Map<String, dynamic>) {
+      throw ApiException(
+          'unable to parse GraphQLResponse from server response which was not a JSON object.',
+          underlyingException: graphQLResponse.body);
+    }
+
+    final responseData = responseBody['data'];
+    // Preserve `null`. json.encode(null) returns "null" not `null`
+    final responseDataJson =
+        responseData != null ? json.encode(responseData) : null;
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/send_graphql_request.dart:47]: callout (no change needed in this PR): Looks like we are decoding the response from string to map, then re-encoding part of it to a string to pass to decoder, which then converts it back to JSON again. Looks like we are doing this bc we are basing it off the decoder which was built around the method channel implementation that returns a string. We should look into changing the decoder to accept a map so that we only have to decode it once.

**Review by ragingsquirrel3 — APPROVED** [review]

File: packages/api/amplify_api/test/amplify_api_config_test.dart (hunk 32 above)
Comment (Equartey) [reply packages/api/amplify_api/test/amplify_api_config_test.dart]: Think you're right, this test isn't needed. It's not invalid per se, and it's not a unique test case.

File: packages/api/amplify_api/test/dart_graphql_test.dart [hunk 34]
```diff
@@ -0,0 +1,216 @@
+import 'dart:convert';
```
Comment (dnys1) [inline packages/api/amplify_api/test/dart_graphql_test.dart:1]: Missing license header on this file. Hoping to create an `aft` command soon but in the meantime you can use `license.sh` in the repo to automatically add them in a package. So from the API package root, you could run `../../../tool/license.sh`

File: packages/api/amplify_api/lib/src/util.dart [hunk 35]
```diff
@@ -30,3 +30,21 @@ Map<String, String>? addContentTypeToHeaders(
   modifiedHeaders.putIfAbsent(AWSHeaders.contentType, () => contentType);
   return modifiedHeaders;
 }
+
+/// Grabs errors from GraphQL Response. Is used in method channel and Dart first code.
+/// TODO: Move to Dart first code when method channel GraphQL implementation is removed.
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/util.dart:35]: TODOs shouldn't be doc comments generally, and lets add usernames for all TODOs (this is a lint in the new packages, but this one hasn't been migrated yet)

```suggestion
// TODO(Equartey): Move to Dart first code when method channel GraphQL implementation is removed.
```

File: packages/api/amplify_api/lib/src/graphql/send_graphql_request.dart [hunk 36]
```diff
@@ -0,0 +1,57 @@
...
+/// Converts the [GraphQLRequest] to an HTTP POST request and sends with ///[client].
+@internal
+Future<GraphQLResponse<T>> sendGraphQLRequest<T>({
+  required GraphQLRequest<T> request,
+  required http.Client client,
+  required Uri uri,
+}) async {
+  try {
+    final body = {'variables': request.variables, 'query': request.document};
+    final graphQLResponse = await client.post(uri, body: json.encode(body));
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/send_graphql_request.dart:34]: Need to check for `200` status code here and throw otherwise

File: packages/api/amplify_api/test/dart_graphql_test.dart [hunk 37]
```diff
@@ -0,0 +1,216 @@
...
+          'id': 'TEST_ID',
+          'name': 'Test App Blog',
+          'createdAt': '2022-06-28T17:36:52.460Z'
+        }
+      ]
+    }
+  }
+};
+
+final _modelQueryId = UUID.getUUID();
```
Comment (dnys1) [inline packages/api/amplify_api/test/dart_graphql_test.dart:32]: Use `uuid` from `aws_common`. We're gonna deprecate this soon.

File: packages/api/amplify_api/test/amplify_api_config_test.dart [hunk 38]
```diff
@@ -0,0 +1,75 @@
...
+          endpoint: endpoint,
+          region: region,
+          authorizationType: authorizationType,
+          apiKey: apiKey);
+
+      endpointConfig = const EndpointConfig('GraphQL', config);
+    });
+
+    test('should return valid URI with null params', () async {
+      final uri = endpointConfig.getUri(path: null, queryParameters: null);
```
Comment (dnys1) [inline packages/api/amplify_api/test/amplify_api_config_test.dart:32]:
```suggestion
      final uri = endpointConfig.getUri();
```

**Review by dnys1 — APPROVED** [review]
Some minor comments. LGTM, though!
